### DISCUSSION TITLE
How to remove trailing "." for supplement with heading numbering function?

### DISCUSSION THREAD

**Discussion by Otto-AA** [typst/typst#4574] [Q&A]
Hi, I'm considering to switch to Typst from Latex and encountered following issue:

When using heading with a numbering function, the supplement contains a trailing ".". For instance:

```typst
#set heading(numbering: "1.")

= Pizza <pizza>

In @pizza, we show a pizza recipe.

#set heading(numbering: (..numbers) => numbering("1.", ..numbers))

= Pasta <pasta>

In @pasta, we show a pasta recipe.
```

Will render as:

```
1. Pizza
In Section 1, we show a pizza recipe.

2. Pasta
In Section 2., we show a pasta recipe.
```

Where the second line contains a trailing ".". Is this a bug, or how could I modify it to get rid of this dot?

If it is relevant, I am using a function to only show the numbering for the first 3 levels (1.2.3), but not deeper (no section 1.2.3.4.5).

**Comment by lars-reimann** [comment]
Instead of the format string, you may also use a custom function:

```
numbering(
  (..nums) => nums
      .pos()
      .map(str)
      .join("."),
  ..numbers
)
```

**Comment by danlyng** [comment]
To restate the problem, there is not an ergonomic way of placing a suffix in a numbering function. If you "manually" add a suffix by doing something like 

```
numbering(
  (..nums) => nums
      .pos()
      .map(str)
      .join(".") + ".",
  ..numbers
)
```
then inline references to the heading will include the period. Without it, the heading will not show a period between the numbering and the body.

This can be solved by forgoing the suffix in the numbering function and adding a show rule like this, but perhaps there can be a more natural solution.
```
#show heading: it => [
  #counter(heading).display(it.numbering). #it.body
  
]
```

**Comment by eike** [comment]
This is a bit off topic, but the entire supplement thing is one of my least favorite things about Typst. I really wonder why there is not simply a way to specify a function that returns the entire reference text. (So for a section heading, you’d give a function returning `"Section 1.2"` instead of a function that returns `"Section"` with Typst adding the 1.2 for you. Lots of these “How do I adapt the numbering” questions would become easier.
